### ISSUE TITLE
As a member of the public (permittee), I can enter in basic information (email) so that I can get a permit for Arapaho and Roosevelt National Forests

### ISSUE THREAD

**Issue by sadlerw** [nciinc/fs-permit-platform#12]
##Acceptance Criteria

- [x] user can enter information that the forest service requires from the user to obtain a permit.
- [x] user should be able to enter a quantity
- [x] user should see the price per permit/tag
- [x] user knows which forest they are entering information for.
- [x] user should be able to submit the information and have it stored in the db for later retrieval.
- [x] user should see verbiage describing what will happen in the flow after submission and that they will be returned to receive the permit. (will be fleshed out later once pay.gov integration is complete)

## Definition of Done
- [x] Code unit / e2e tested and comply with QWASP
- [x] No new security vulnerabilities
- [x] Internal team code review
- [x] Documentation / readme.md / CircleCI updated
- [x] Docker updated
- [x] Compare finished design with mockup
- [x] Usability tested (Bryan)

## Tasks

- [x] research and get pricing information for each pilot forest (Jane)
- [x] user task flow model (Bryan, Jane)
- [x] db model (email address, cost per tree and forest org id) (Shekar)
- [x] migration (use dotransaction) (Shekar)
- [x] routes in frontend (Cameron, Seth)
- [x] write up form help text and boilerplate (Jane)
- [x] html design mockup (Cameron, Seth)
- [x] design approval (Melissa, Amber)
- [x] view/component (Cameron, Seth)
- [x] REST endpoints (GET POST)
- [x] simplify/remove e2e functions that are used only once

**Comment by ASprinkle** [comment]
I would like this story to be focused on the is  the entering of the permittee information (before you pay)

**Comment by MelissaBraxton** [comment]
Here is a rough user task flow I sketched out during our chat. its a first take, and I suspect a more ideal user experience that still meets the overall payment goal could be achieved in one or two fewer tasks, but we'll see!
![xmas_tree_payment_userflow](https://user-images.githubusercontent.com/20651166/33146064-ad393108-cf77-11e7-9a97-9848e893ca8c.JPG)

**Comment by ASprinkle** [comment]
@JaneZC I see you're task is research to determine pricing information. Here is some information that was emailed to me recently from the Forest contacts I have:

**Flathead National Forest:** 
-people are limited to 3 tags per household for the Forest. This limit does not vary by district.
-Christmas tree permits are $5.00 each across the Forest. This price does not vary by district.

**Shoshone National Forest**
-Christmas tree permits are $8.00 per tree if the tree is less than 20 feet tall.  If the tree is over 20 feet tall then it's an additional $8.00 up to 30 feet, and then an additional $8.00 for each additional 10 feet.  This pricing structure is consistent across the Forest.
-people are limited to 5 trees per household. This is consistent across the Forest.

**Mt. Hood National Forest**
-Christmas tree permits are $5.00 per tree. This pricing structure is the same across forests
-There is a limit of five trees per household.  This is the same across the forest.

I haven't heard back from the ARNF yet

**Comment by JaneZC** [comment]
@adsprinkle Thank you!!

A question on terminology - are 'tags' and 'permits' interchangeable? So, for Flathead, you'd buy three permits and get three tags? I suspect it would be best to only user the word 'permit' on the form, but wanted to make sure I understand how that terminology is used.

Arapaho & Roosevelt National Forests 

- Permits cost $10 per tree 
- Permits are limited to a maximum of five per person.
- It appears that the pricing is consistent across the Forest Districts; ARNF has posted their 2017 information online.

**Comment by sethalt** [comment]
Attached are screenshots of the basic form application layout. One screen shot is with the fields filled in without errors, the other show the error messages. Lets start thinking about form hint text or error messaging. @JaneZC @bryanmcfadden @MelissaBraxton 

![screen shot 2017-11-27 at 10 28 04 am](https://user-images.githubusercontent.com/2347759/33277450-3060b90e-d35e-11e7-9662-55457df3b877.png)

![screen shot 2017-11-27 at 10 26 31 am](https://user-images.githubusercontent.com/2347759/33277457-3b851e38-d35e-11e7-9ea5-bd2bd0d8e3a8.png)

**Comment by sethalt** [comment]
@JaneZC @bryanmcfadden @MelissaBraxton 

The language for redirecting to pay.gov. This current is the page that appears when you click "Submit your application," however it can be put inline on the form page or appear as a popup if you all think that is better. From a 508 standpoint, I think the text alone on a new screen might be the easiest/best solution.

![screen shot 2017-11-27 at 11 28 29 am](https://user-images.githubusercontent.com/2347759/33280107-992a6cc0-d366-11e7-960e-e9d3d721398d.png)


On the intake side, handling the login.gov redirect notice, we added text right in the button that told the user they would login with Login.gov. E.g. 
![screen shot 2017-11-27 at 11 31 29 am](https://user-images.githubusercontent.com/2347759/33280146-bd22e7f6-d366-11e7-9c98-2a6c0b9b0371.png)

**Comment by ASprinkle** [comment]
@JaneZC and team - let's stick with the word "permit" vs. tags. In the office it seems like they are interchangeable - but for purposes of this project let's be consistent in using the term "permit".

The Apply form looks great - for the "Number of trees" field, can we title it "Number of Christmas tree permits to purchase" or something like that?

**Comment by sethalt** [comment]
@JaneZC @bryanmcfadden @MelissaBraxton @ASprinkle Updated `confirm and pay` page
![screen shot 2017-11-27 at 3 16 01 pm](https://user-images.githubusercontent.com/2347759/33289698-045d0c72-d386-11e7-869c-9b7b61047270.png)

**Comment by ASprinkle** [comment]
I think that looks great! The only comments I have are:
-Change "number of trees" to "Number of Christmas tree permits" or something along those lines.
-Does being directed to pay.gov actually "complete" the transaction?

**Comment by MelissaBraxton** [comment]
Great progress! A few comments:

The checkout button is quite long.  Consider not including the pay.gov logo
inside the button container, but elsewhere on the page.

In this Peace Corps example
<https://www.peacecorps.gov/donate/projects/the-marine-sanctuary-revitalization-project-17-492-013/donate/?amount=50>,
they don't use the pay.gov logo at all. I like the way they address
@asprinkle process question--They tell the user that they're going to pay
on pay.gov and but that they'll be directed back to the peace corps site,
which is the same flow (I believe) users will experience with xmas tree
permitting.


Melissa Braxton
User Experience Designer, 18F
Cell: (415) 696-4216
https://18f.gsa.gov/

**Comment by JaneZC** [comment]
A few suggestions .... @adsprinkle @MelissaBraxton @bryanmcfadden 

1) I would recommend user testing the header (or we can change now?) It may be more clear to use "Buy a Christmas tree permit" instead of "Apply for a Christmas tree permit" ... This page doesn't mention to the user that the goal is for them to purchase permit ... just asks them to apply for a permit, gives them a price, and tells them to 'submit' their application. Using the word 'buy' in the header may clarify things a bit.

2) @adsprinkle ... I agree with Number of Christmas tree permits as wording for the field. 

3) In the future, can we add how much the tree costs to the page itself? I think it would be nice to have that information up front ... instead of having to backwards calculate it from the form. 

5) It is Arapaho and Roosevelt National Forests (with an s). Also, Christmas should be capitalized in the header. 

6) The Confirm and Pay page should use the word permit on it somewhere. (Right now, it sounds like you are ordering trees instead of permits.) Is it possible to say Number of Christmas tree permits here as well ... instead of trees?

**Comment by JaneZC** [comment]
@sethalt @bryanmcfadden 

Below are some suggested error messages; I wrote them to focus on what the user could do to fix the message (instead of telling them what they probably have figured out.) Permit holder is a bit long ... but I think a necessary evil for the initial go around. :)

Suggested Error Messages: 

Please enter the permit holder's first name.

Please enter the permit holder's last name.

Please enter an email address. (if nothing is entered)
Please enter a valid email address. (if an invalid email is entered)

Please enter the number of Christmas trees.

Also, like @MelissaBraxton , I think the the Peace Corp pay.gov message is clean and simple. Can we add to the page with a minor change? (We can leave off the word NOTE though.) So, just the below:

We use Pay.gov as our trusted secure payment processor. We’ll send you on to there to buy your permit(s) and bring you back when you finish.

**Comment by sethalt** [comment]
@ASprinkle @MelissaBraxton, I've updated the views with the feedback provided. @JaneZC I added some intro text with price per tree and number of trees per household info. Please revise if needed.

<img width="690" alt="screen shot 2017-11-27 at 7 04 15 pm" src="https://user-images.githubusercontent.com/2347759/33297373-3fdc1688-d3a6-11e7-855c-83584ca75ce1.png">

<img width="865" alt="screen shot 2017-11-27 at 7 04 25 pm" src="https://user-images.githubusercontent.com/2347759/33297376-452e2360-d3a6-11e7-8863-5907ec80e06c.png">

**Comment by sethalt** [comment]
@JaneZC Regarding the error messages. We are using the paradigm established on the intake project, and I'm hesitant drifting from it unless there is strong evidence that we need to. We are using a re-usable error template so changing the messaging would require a new error template just for the trees form fields.

**Comment by sethalt** [comment]
@ASprinkle @JaneZC what about adding some hint text below the email label, so a user doesn't need to tab past the input to know why the field is necessary?

**Comment by ASprinkle** [comment]
That sounds even better!

**Comment by JaneZC** [comment]
@sethalt @adsprinkle

For email hint text .... Hint text should help users fill out the form fields with proper format (ie. passwords, phone numbers, etc.) I am not sure we need it as I think email format is common enough that the error that will occur is largely due to typos and/or mistakes. But, I don't think it hurts anything. If you'd like to add it, I would use this (similar to what is on the DOB input here - https://standards.usa.gov/components/form-controls/#date-input

For example: someone@example.com

One additional thought ... do we need a first name and last name field or can we just say Name? I am not sure the database is built or what the requirements would be for using that data elsewhere (ie on print-outs, etc.), but was thinking it might be easier on users by having less fields for them to fill out.

**Comment by JaneZC** [comment]
@sethalt Agreed on error messages. It should be consistent. Thanks for sharing!!!

**Comment by sethalt** [comment]
@MelissaBraxton @ASprinkle 

The team decided that the confirm page wasn't necessary at this point, so we condensed it down to a single page. We've updated the copy for clarity. Please let us know what you think.

![screen shot 2017-11-28 at 9 36 30 am](https://user-images.githubusercontent.com/2347759/33328480-b76d2e84-d41f-11e7-89ab-ba8714f0b237.png)

**Comment by sethalt** [comment]
and here it is with the fields filled in
![screen shot 2017-11-28 at 9 39 04 am](https://user-images.githubusercontent.com/2347759/33328591-06a8a6cc-d420-11e7-8b94-62b0ef93291f.png)

**Comment by ASprinkle** [comment]
@sethalt That looks really clear to me. 
@JaneZC First and Last name is preferred over "name".  Their first name/last name, or some combination will be printed out on the permit somewhere.

**Comment by sethalt** [comment]
@ASprinkle @MelissaBraxton please comment with `Approved` if you feel like this design meets the acceptance criteria, and if so we will merge this code into sprint-4. Thanks!

**Comment by MelissaBraxton** [comment]
Looks good to me!  I checked the "approved" box! :) 

A few minor tweaks to consider in the future:
 - Add a line break between "Christmas tree permits cost..." and Your Information to make it clear that the "Your info" header is tied to the form fields below, not the section above.
 - Remove the period after "Buy a Christmas Tree Permit" (I know it is an imperative, but it is inconsistent with the header style elsewhere.)

**Comment by sethalt** [comment]
@MelissaBraxton there is a line break, but it must have gotten lost when i zoomed out to take the screen shot. See below. 

<img width="593" alt="screen shot 2017-11-28 at 2 21 20 pm" src="https://user-images.githubusercontent.com/2347759/33342375-7bca4146-d447-11e7-800f-2a63a897d9b6.png">

As for the period. I'd like to discuss that with you. That paradigm was carried over from Intake (see below), and perhaps we need to re-address how we use periods in headings on our tree guidelines page.

![screen shot 2017-11-28 at 2 23 11 pm](https://user-images.githubusercontent.com/2347759/33342478-d417c1a2-d447-11e7-8775-233a06797400.png)

**Comment by MelissaBraxton** [comment]
Thanks @sethalt 
 - I am checking on the period issue w/ our content folks. Our content guide is silent on the matter. I'd leave the period in for now to stay consistent with the special uses site style for now. 
 - One more thing! Please shorten the length of the "number of Christmas tree permits" form field. Since the limit is 5, people's answer will only be a single character.

**Comment by sethalt** [comment]
Pre-merge screenshot with some of the last recommendations.

![screen shot 2017-11-28 at 4 00 44 pm](https://user-images.githubusercontent.com/2347759/33346495-8efbead6-d455-11e7-86a8-4e30939acd08.png)

**Comment by bryanmcfadden** [comment]
![user_task_flow_permit_purchase_draft_01](https://user-images.githubusercontent.com/16689692/33347694-bf0deb46-d461-11e7-83b7-c5d26c922ea3.jpg)

Based off of Melissa's design. 

A few questions:
1. Do we need a flow for pay.gov (I think its needed for the next story)?
2. How does pay.gov work? Does it provide a summary detail?
3. Where does the user end up if they cancel the redirect to pay.gov?

**Comment by ASprinkle** [comment]
These are great questions, and a bit beyond my expertise.. I will defer to others on the team.

**Comment by lpendem** [comment]
@ASprinkle @MelissaBraxton Direct URL to the permit application form for Mt. Hood is : https://forest-service-trees-staging.app.cloud.gov/applications/christmas-trees/forests/mthood/new . Replace <mthood> in the URL with <arp> or <flathead> for other forests.

**Comment by MelissaBraxton** [comment]
@bryanmcfadden and @JaneZC 

Some thoughts on your questions:

1. I think it's fine to flesh out the task and info flow for pay.gov as part of working on #13 
2. Another thing to be worked out fully in #13, but it does have implications for this flow diagram. Regardless of whether pay.gov provides a summary detail, I'd want to see a purchase confirmation from the FS. From what I've heard re: the login.gov experience (not analogous, I know), being redirected to fed gov. a site outside the one they intended to visit can be confusing for folks. I'd imagine people getting worried if the order confirmation/summary did not come from the FS, but if this is how pay.gov works, then we'll just need to see what we can do about the design of the summary detail page pay.gov provides (e.g., include the FS logo).
3. I'm not sure there needs to be a separate step to accept or cancel the pay.gov redirect. Seems fine to me to have clicking "checkout" go right to the payment form. If folks want to cancel their transaction from their, I'd imagine going back to the buy a permit screen, rather than the forest-specific page. 

Another Q for the team:

**Comment by lauraGgit** [comment]
One minor thing - I think it could be helpful in the number of christmas tree help text to include the maximum number there too.

**Comment by sethalt** [comment]
@lauraGgit @MelissaBraxton @ASprinkle 

![screen shot 2017-12-01 at 7 54 33 am](https://user-images.githubusercontent.com/2347759/33485685-eccd231e-d66c-11e7-9f09-1c3e4fee4010.png)

**Comment by JaneZC** [comment]
@MelissaBraxton @sethalt 

I wanted to check and see what was the final decision was on whether a period is needed after 'buy a permit'? 

Thank you!

**Comment by MelissaBraxton** [comment]
Thanks @JaneZC! 

Following up re:  "." at the end of headings that are full sentences. 

I checked in with our content folks, and the verdict is no "." 

I know this is different than what had been decided on the intake module, (we do add periods at the end of full sentence list items, and our content folks wondered whether the `li` guidance got conflated with heading style.)  Also, the APA style guide does say that a period is ok at the end of a full sentence heading, but only lower level headings, not H1 or H2.
